### ISSUE TITLE
Installation issue: fd@9.0.0 on macOS using Rust from develop-ml-darwin-aarch64-mps buildcache

### ISSUE THREAD

**Issue by pranav-sivaraman** [LLNL/spack#43854]
### Steps to reproduce the issue

```console
$ spack spec -I fd
Input spec
--------------------------------
 -   fd

Concretized
--------------------------------
 -   fd@9.0.0%apple-clang@15.0.0 build_system=cargo arch=darwin-sonoma-m1
[+]      ^rust@1.76.0%apple-clang@15.0.0~dev~docs+src build_system=generic arch=darwin-ventura-aarch64
[+]          ^cmake@3.27.9%apple-clang@15.0.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=darwin-ventura-aarch64
[+]              ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-ventura-aarch64
[+]              ^ncurses@6.4%apple-clang@15.0.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=darwin-ventura-aarch64
[+]              ^zlib-ng@2.1.6%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-ventura-aarch64
[+]          ^curl@8.6.0%apple-clang@15.0.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=secure_transport arch=darwin-ventura-aarch64
[+]              ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-aarch64
[+]              ^nghttp2@1.57.0%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-aarch64
[+]          ^libgit2@1.7.2%apple-clang@15.0.0~curl~ipo+mmap+ssh build_system=cmake build_type=Release generator=make https=system arch=darwin-ventura-aarch64
[+]              ^libssh2@1.11.0%apple-clang@15.0.0+shared build_system=autotools crypto=openssl patches=011d926 arch=darwin-ventura-aarch64
[+]              ^pcre@8.45%apple-clang@15.0.0~jit+multibyte+pic+shared+static+utf build_system=autotools arch=darwin-ventura-aarch64
[+]          ^ninja@1.11.1%apple-clang@15.0.0+re2c build_system=generic arch=darwin-ventura-aarch64
[+]              ^re2c@2.2%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-aarch64
[+]          ^openssl@3.2.1%apple-clang@15.0.0~docs+shared build_system=generic certs=mozilla arch=darwin-ventura-aarch64
[+]              ^ca-certificates-mozilla@2023-05-30%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-aarch64
[+]              ^perl@5.38.0%apple-clang@15.0.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=darwin-ventura-aarch64
[+]                  ^berkeley-db@18.1.40%apple-clang@15.0.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=darwin-ventura-aarch64
[+]          ^pkgconf@1.9.5%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-aarch64
[+]          ^python@3.11.7%apple-clang@15.0.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=darwin-ventura-aarch64
[e]              ^apple-libuuid@1353.100.2%apple-clang@15.0.0 build_system=bundle arch=darwin-ventura-aarch64
[+]              ^bzip2@1.0.8%apple-clang@15.0.0~debug~pic+shared build_system=generic arch=darwin-ventura-aarch64
[+]                  ^diffutils@3.10%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-aarch64
[+]              ^expat@2.6.2%apple-clang@15.0.0~libbsd build_system=autotools arch=darwin-ventura-aarch64
[+]              ^gdbm@1.23%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-aarch64
[+]              ^gettext@0.22.4%apple-clang@15.0.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=darwin-ventura-aarch64
[+]                  ^libiconv@1.17%apple-clang@15.0.0 build_system=autotools libs=shared,static arch=darwin-ventura-aarch64
[+]                  ^libxml2@2.10.3%apple-clang@15.0.0+pic~python+shared build_system=autotools arch=darwin-ventura-aarch64
[+]                  ^tar@1.34%apple-clang@15.0.0 build_system=autotools zip=pigz arch=darwin-ventura-aarch64
[+]                      ^pigz@2.8%apple-clang@15.0.0 build_system=makefile arch=darwin-ventura-aarch64
[+]                      ^zstd@1.5.6%apple-clang@15.0.0+programs build_system=makefile compression=none libs=shared,static arch=darwin-ventura-aarch64
[+]              ^libffi@3.4.6%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-aarch64
[+]              ^libxcrypt@4.4.35%apple-clang@15.0.0~obsolete_api build_system=autotools patches=4885da3 arch=darwin-ventura-aarch64
[+]              ^readline@8.2%apple-clang@15.0.0 build_system=autotools patches=bbf97f1 arch=darwin-ventura-aarch64
[+]              ^sqlite@3.43.2%apple-clang@15.0.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=darwin-ventura-aarch64
[+]              ^xz@5.4.6%apple-clang@15.0.0~pic build_system=autotools libs=shared,static arch=darwin-ventura-aarch64
[+]          ^rust-bootstrap@1.75.0%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-aarch64


```

### Error message

<details><summary>Error message</summary>

<pre>
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/nghttp2-1.57.0-kiaime2am52ys5pkmqvnnmvwvbwdtxag
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/zlib-ng-2.1.6-gupz2crim2jetppx3b5jawmlyo57np4b
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/xz-5.4.6-nrzvy3emno3sqpjnhppe2xfq7okvzyml
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/pcre-8.45-lpid4y7l7fka4iu2t5no52o47g7swszr
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/openssl-3.2.1-cakaxzjmk3mmng2d2hdbbrqd6x4z5bgc
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/curl-8.6.0-ocnkqfqnfqywzsmy2kvjoqnukqfzlons
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/libssh2-1.11.0-iqlkhqgjruufdv6bms3u6insejhou2oz
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/libgit2-1.7.2-ruyax7pes35jbgswlh6z3cqvubc6layi
[+] /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/rust-1.76.0-yvcqz7rm6h4uugqh2reykv72vz6kvoao
==> Installing fd-9.0.0-3vlt565ioy33uwdrx7rlqzoqgjowz77m [10/10]
==> No binary for fd-9.0.0-3vlt565ioy33uwdrx7rlqzoqgjowz77m found: installing from source
==> Using cached archive: /Users/pranav/spack/var/spack/cache/_source-cache/archive/30/306d7662994e06e23d25587246fa3fb1f528579e42a84f5128e75feec635a370.tar.gz
==> No patches needed for fd
==> fd: Executing phase: 'build'
==> [2024-04-25-14:24:26.667165] '/Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/rust-1.76.0-yvcqz7rm6h4uugqh2reykv72vz6kvoao/bin/cargo' 'install' '--root' 'out' '--path' '.'
dyld[20497]: Library not loaded: /opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib
  Referenced from: <24EDA3A0-6C05-3F6C-BD25-8F2B3B3C66C0> /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/rust-1.76.0-yvcqz7rm6h4uugqh2reykv72vz6kvoao/bin/cargo
  Reason: tried: '/opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib' (no such file), '/opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib' (no such file)
==> Error: ProcessError: Command exited with status -6:
    '/Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/rust-1.76.0-yvcqz7rm6h4uugqh2reykv72vz6kvoao/bin/cargo' 'install' '--root' 'out' '--path' '.'
See build log for details:
  /var/folders/0n/cq9bc7093psd1kqz6c69lrhc0000gn/T/pranav/spack-stage/spack-stage-fd-9.0.0-3vlt565ioy33uwdrx7rlqzoqgjowz77m/spack-build-out.txt

</pre></details>

### Information on your system

* **Spack:** 0.22.0.dev0 (4a40a76291f05fdd76e12497b1891c5ad1705ebe)
* **Python:** 3.9.6
* **Platform:** darwin-sonoma-m1
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/15119369/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/15119370/spack-build-out.txt)
@alecbcs @ashermancinelli 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by alecbcs** [comment]
@pranav-sivaraman it looks like you're using an external `openssl` that was installed with brew. If you disable that external and instead use a Spack built `openssl` is fd successful?

**Comment by pranav-sivaraman** [comment]
@alecbcs I am using the `openssl` from the Spack buildcache. I have no externals defined in `packages.yaml`. I got rid of `brew` and I am trying to create my devtools stack in Spack. Unless there are still some lingering stuff from brew which I don't think so since `/opt/homebrew/` is empty. Maybe I need to build `openssl` locally instead of from the buildcache?

**Comment by alecbcs** [comment]
Try building all the way up locally and see if that fixes it. We may have a bug in the public MacOS cache. I know there was some discussion about a bug in binary relocation on MacOS a while back. This may be a good reproducer if true.

**Comment by pranav-sivaraman** [comment]
Yep building `rust` from source seems to fix the issue. Most dependencies are from the public MacOS cache.

**Comment by pranav-sivaraman** [comment]
Here is the output of just running `cargo` from the buildcache and also the list of libraries. 
```shell
./cargo
dyld[34329]: Library not loaded: /opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib
  Referenced from: <D1E6AEE9-AEFD-30FF-A7D1-842217BBB562> /Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/rust-1.76.0-hba2bhav6rzc7dazmmwzlsmw2bizrjlc/bin/cargo
  Reason: tried: '/opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib' (no such file), '/opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib' (no such file), './libssl.1.1.dylib' (no such file)

otool -L ./cargo
cargo:
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 61040.61.1)
	@rpath/libgit2.1.7.dylib (compatibility version 1.7.0, current version 1.7.2)
	/opt/homebrew/opt/openssl@1.1/lib/libssl.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	/opt/homebrew/opt/openssl@1.1/lib/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	@rpath/libz.1.dylib (compatibility version 1.0.0, current version 1.3.0)
	/Users/pranav/spack/opt/spack/darwin-ventura-aarch64/apple-clang-15.0.0/curl-8.6.0-ocnkqfqnfqywzsmy2kvjoqnukqfzlons/lib/libcurl.4.dylib (compatibility version 13.0.0, current version 13.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 2202.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)

```

**Comment by pranav-sivaraman** [comment]
This seems to have been fixed with the latest version (`rust@1.78`). It relocated fine from the buildcache and now I can install cargo packages just fine.
